### DISCUSSION TITLE
line skipping and oversized bounding boxes

### DISCUSSION THREAD

**Discussion by robpetrosino** [PaddlePaddle/PaddleOCR#13640] [Q&A] [answered]
I am trying to understand why in some cases PaddleOCR fails to identify some text lines as such, and therefore skip them in the text recognition algorithm.  Here's an example:

![ocr_IMG_4216_2](https://github.com/user-attachments/assets/409ee8a8-5c91-46bc-a873-37f8fed408d2)

One line in the middle and the last one are skipped, in the sense that they seem to have been skipped in the text detection procedure altogether. 

Sometime it also happens that a whole paragraph is being skipped, and some boxes are bigger than the others, thus overlapping with the boxes of the following lines: 

![ocr_IMG_4219_1](https://github.com/user-attachments/assets/07d6ace6-3927-4c10-bcbe-0bfd602689c9)

Can anyone suggest how to address the problem? I have been playing with the hyperparameters of the model but I can't seem to make sensible progress. Here's a snippet of my code:

```
from paddleocr import PaddleOCR, draw_ocr
import glob
import re 
import cv2

ocr = PaddleOCR(use_angle_cls=True, lang='en',
                det_db_thresh = 0.4, # only pixels with a score greater than this threshold will be considered as text pixels
                det_db_box_thresh = 0.4, # when the average score of all pixels is greater than the threshold, the result will be considered as a text area
                det_db_unclip_ratio = 1.6, # the expansion factor of the Vatti clipping algorithm, which is used to expand the text area
                max_batch_size = 32,
                det_limit_side_len = 1000, # image side length limit
                det_db_score_mode = "fast", # DB detection result score calculation method, supports fast and slow, fast calculates the average score according to all pixels within the bounding rectangle of the polygon, slow calculates the average score according to all pixels within the original polygon, The calculation speed is relatively slower, but more accurate.
                dilation = True, # Whether to inflate the segmentation results to obtain better detection results
                ocr_version = "PP-OCRv4") 

def save_ocr(img_path, show=False):
  page_name = re.findall('cut/(.+).png', img_path)[0] # save the scan name
  print(f"\nIMG: {page_name}\n")
  save_path = './ocr_output/ocr_' + page_name + '.png'

  result = ocr.ocr(img_path)
  boxes = [] #list of the bounding boxes
  txts = [] #list of all OCR'ed lines
  scores = [] #list of the per-line confidence score
  for lines in result:
    for line in lines:
      boxes.append(line[0])
      txts.append(line[1][0])
      scores.append(line[1][1])

  img = cv2.imread(img_path)
  im_show = draw_ocr(img, boxes, txts, scores, font_path='./fonts/simfang.ttf')

  cv2.imwrite(save_path, im_show)

  if show:
    img = cv2.cvtColor(im_show, cv2.COLOR_BGR2RGB)
    plt.imshow(img)
  
  return boxes, txts, scores

scans_path = sorted(glob.glob('./cut/*.png'))

all_txts = [] #list of all OCR'ed lines of all scans
for scan in scans_path:
  _, txts, _ = save_ocr(scan)
  all_txts.append(txts)

```

Thank you all in advance!

**Comment by UserWangZz** [comment]
You can find the meaning of relevant hyperparameters from this document
https://paddlepaddle.github.io/PaddleOCR/ppocr/blog/inference_args.html
Here I recommend you try the following hyperparameter combinations
det_db_box_thresh defaults to 0.6, you can try to lower it
det_db_unclip_ratio defaults to 1.5, you can try to increase it
use_dilation defaults to False, you can try setting it to true
det_db_score_mode defaults to fast, you can try setting it to slow. I have a scenario very similar to your second picture. Maybe this parameter can solve your problem well.

**Reply by robpetrosino** [reply]
Thanks for your reply. Yeah -- I tried to play a bit with these parameters (and the others listed in the code above). It looks like `det_db_score_mode` increases the chances of a line to get detected correctly, though some lines are still being skipped altogether. 

I noticed that this happens only on the left-sided pages (where the lines may be "curved" due to binding). Do you have any suggestion of how I could "flatten" these lines?

Thank you in advance!

**Reply by UserWangZz** [reply, marked as answer]
det_db_score_mode
Try setting this parameter to slow. Will it help this problem? Because I also encountered the situation of row bending. After setting this parameter, it can be detected, although the detection frame cannot be fully covered because of the bending.

**Reply by robpetrosino** [reply]
Yes -- that actually makes things a bit better, though some lines are still being skipped (see below), even after performing some morphological transformation (erosion + dilation). 

![ocr_IMG_4219_1](https://github.com/user-attachments/assets/b118f467-4d16-43fd-9472-6aa8196dc8e3)

So I guess it all boils down to figure out how to undo the character bending... 

Thanks!
